lib/libc/gen/popen.3
.\"	$NetBSD: popen.3,v 1.25 2022/12/04 11:25:08 uwe Exp $
.\"
.\" Copyright (c) 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)popen.3	8.2 (Berkeley) 5/3/95
.\"
.Dd September 11, 2021
.Dt POPEN 3
.Os
.Sh NAME
.Nm popen ,
.Nm popenve ,
.Nm pclose
.Nd process
.Tn I/O
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In stdio.h
.Ft FILE *
.Fn popen "const char *command" "const char *type"
.Ft FILE *
.Fn popenve "const char *path" "char * const *argv" "char * const *envp" "const char *type"
.Ft int
.Fn pclose "FILE *stream"
.Sh DESCRIPTION
The
.Fn popen
function
.Dq opens
a process by creating an IPC connection,
forking,
and invoking the shell.
Historically,
.Fn popen
was implemented with a unidirectional pipe;
hence many implementations of
.Fn popen
only allow the
.Fa type
argument to specify reading or writing, not both.
Since
.Fn popen
is now implemented using sockets, the
.Fa type
may request a bidirectional data flow.
The
.Fa type
argument is a pointer to a null-terminated string
which must be
.Ql r
for reading,
.Ql w
for writing, or
.Ql r+
for reading and writing.
In addition if the character
.Ql e
is present in the
.Fa type
string, the file descriptor used internally is set to be closed on
.Xr exec 3 .
.Pp
The
.Fa command
argument is a pointer to a null-terminated string
containing a shell command line.
This command is passed to
.Pa /bin/sh
using the
.Fl c
flag; interpretation, if any, is performed by the shell.
.Pp
The
.Fn popenve
function is similar to
.Fn popen
but the first three arguments are passed
to
.Xr execve 2
and there is no shell involved in the command invocation.
.Pp
The return value from
.Fn popen
and
.Fn popenve
is a normal standard
.Tn I/O
stream in all respects
save that it must be closed with
.Fn pclose
rather than
.Fn fclose .
Writing to such a stream
writes to the standard input of the command;
the command's standard output is the same as that of the process that called
.Fn popen ,
unless this is altered by the command itself.
Conversely, reading from a
.Dq popened
stream reads the command's standard output, and
the command's standard input is the same as that of the process that called
.Fn popen .
.Pp
Note that output
.Fn popen
streams are fully buffered by default.
.Pp
The
.Fn pclose
function waits for the associated process to terminate
and returns the exit status of the command
as returned by
.Xr wait4 2 .
.Sh RETURN VALUES
The
.Fn popen
function returns
.Dv NULL
if the
.Xr vfork 2 ,
.Xr pipe 2 ,
or
.Xr socketpair 2
calls fail,
or if it cannot allocate memory, preserving
the errno from those functions.
.Pp
The
.Fn pclose
function
returns \-1 if
.Fa stream
is not associated with a
.Dq popened
command, if
.Fa stream
has already been
.Dq pclosed ,
setting errno to
.Er ESRCH ,
or if
.Xr wait4 2
returns an error, preserving the errno returned by
.Xr wait4 2 .
.Sh SEE ALSO
.Xr sh 1 ,
.Xr execve 2 ,
.Xr fork 2 ,
.Xr pipe 2 ,
.Xr socketpair 2 ,
.Xr vfork 2 ,
.Xr wait4 2 ,
.Xr fclose 3 ,
.Xr fflush 3 ,
.Xr fopen 3 ,
.Xr shquote 3 ,
.Xr stdio 3 ,
.Xr system 3
.Sh STANDARDS
The
.Fn popen
and
.Fn pclose
functions conform to
.St -p1003.2-92 .
.Sh HISTORY
A
.Fn popen
and a
.Fn pclose
function appeared in
.At v7 .
.Pp
The
.Fn popenve
function first appeared in
.Nx 8 .
.Sh BUGS
Since the standard input of a command opened for reading
shares its seek offset with the process that called
.Fn popen ,
if the original process has done a buffered read,
the command's input position may not be as expected.
Similarly, the output from a command opened for writing
may become intermingled with that of the original process.
The latter can be avoided by calling
.Xr fflush 3
before
.Fn popen .
.Pp
Failure to execute the shell
is indistinguishable from the shell's failure to execute the command,
or an immediate exit of the command.
The only hint is an exit status of 127.
.Pp
The
.Fn popen
argument
always calls
.Xr sh 1 ,
never calls
.Xr csh 1 .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.2938 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------

Changes from NetBSD 9.0 to NetBSD 10.0:
	openldap: Import 2.4.48. [christos 20190808]
	boot(8): GPT and RAIDframe support for x86 bootstrap [manu 20190818]
	dhcpcd(8): Import dhcpcd-8.0.3 [roy 20190821]
	inet6: Fix sending RTM_MISS for unresolvable addresses and restore
		ND6_LLINFO_WAITDELETE so that any send at this point can
		return EHOSTDOWN to match inet behaviour. [roy 20190901]
	inet: Fix sending RTM_MISS for unresolvable addresses. [roy 20190901]
	dhcpcd(8): Import dhcpcd-8.0.4 [roy 20190904]
	bind: Import version 9.14.5. [christos 20190905]
	resolvconf(8): Import openresolv-3.9.2 [roy 20190908]
	boot(8): multiboot 2 support, Xen can now boot from EFI [manu 20190913]
	dhcpcd(8): Import dhcpcd-8.0.6 [roy 20190913]
	kernel: Add vHCI, a driver that allows to send and receive USB
		packets from userland. [maxv 20190914]
	mos(4): Ported driver for MosChip MCS7730/7830/7832 USB ethernet.
		[mrg 20190920]
	gcc(1): Import GCC 8.3. [mrg 20191001]
	libpcap: Import 1.9.1. [christos 20191001]
	tcpdump(8): Import 4.9.3. [christos 20191001]
	rescue(8): Moved into a separate, rescue-only set. [maya 20191001]
	kqueue(2): Switch kevent udata from intptr_t void* for compatibility
		with other BSDs. [kamil 20191003]
	byacc: Update to 20190617. [christos 20191006]
	tzdata updates from 2019a to 2019c (incl 2019b) [kre 20191008]
	services(5): Update to version 2019-10-04 from IANA.
		[sevan 20191009]
	dhcpcd(8): Import dhcpcd-8.1.0 [roy 20191011]
	OpenSSH: Import 8.1. [christos 20191012]
	acpi(4): Updated ACPICA to 20190816. [christos 20191015]
	sup: removed, use ftp://ftp.astron.com/pub/sup [christos 20191015]
	dhcpcd(8): Import dhcpcd-8.1.1 [roy 20191016]
	evbarm: Add support for the NanoPi R1 and Duo2. [bad 20191016]
	bind: Import version 9.14.7. [christos 20191017]
	libc: Changed the iconv(3) function signature to the POSIX variation
		[kamil 20191024]
	nct(4): Add Nuvoton NCT5104D GPIO driver. [ad 20191025]
	vio9p(4): Add virtio-9p front-end driver. [ozaki-r 20191028]
	mount_9p(8): Enable to mount a filesystem exported via virtio-9p.
		[ozaki-r 20191028]
	evbarm: Update TI AM335x SoC support to use FDT based configuration.
		[jmcneill 20191028]
	evbarm: Update TI OMAP3 SoC support to use FDT based configuration.
		[jmcneill 20191101]
	amd64: Add support for kCSan - Kernel Concurrency Sanitizer.
		[maxv 20191105]
	audioplay(1): Add support for playing IEEE float32 and float64
		RIFF WAVE files [mrg 20191109]
	tmux(1): Imported 2.9a. [christos 20191112]
	dhcpcd(8): Import 8.1.2. [roy 20191113]
	amd64: Add support for kMSan - Kernel Memory Sanitizer.
		[maxv 20191114]
	kernel: Add fexecve(2) system call for executing a file from a file
		descriptor, conforming to The Open Group Extended API Set 2.
		[christos 20191115]
	qat(4): Add opencrypto driver for Intel QuickAssist.
		Atom C2XXX, C3XXX, Xeon D-21XX and D-15XX, C62X chipsets and
		QuickAssist Adapter 8960/8970.
		[hikaru 20191120]
	select(2), poll(2): Performance and scalability improvements
		[ad 20191121]
	amd64, i386: Performance and scalability improvements for TLB
		shootdown [ad 20191121]
	bind: Import version 9.14.8. [christos 20191127]
	terminfo: Import 20190609 [christos 20191127]
	onewire(4), owtemp(4): Improve reliability and reduce CPU overhead
		slightly. [ad 20191130]
	scheduler: Improve performance and interactivity. [ad 20191201]
	vfs: Reduce false sharing and lock overhead during normal operation.
		[ad 20191201]
	evbarm: Added driver for Allwinner Crypto Engine random number
		generator [riastradh 20191209]
	ixl(4): Ported driver for Intel Ethernet 700 series
		[yamaguchi 20191210]
	acpi(4): Updated ACPICA to 20191213. [christos 20191214]
	uvm: Replace red-black tree page lookup with radix tree lookup,
		including tagging and gang lookup of pages.  Original work
		done by yamt@. [ad 20191214]
	nsd: Import 4.2.4. [christos 20191215]
	unbound(8): Import 1.9.6. [christos 20191215]
	openpam(3): update to 20190224 (tabebuia) [christos 20191215]
 	kerberos(8): Update to Heimdal-7.7.0 [christos 20191215]
	file(1): Upgraded to 5.38. [christos 20191216]
	uvm: Replace global counters with per-CPU counters, significantly
		reducing system time on workloads that are VM-system heavy.
		[ad 20191216]
	anxdp(4): Add driver for Analogix DisplayPort core found in
		Rockchip RK3399. [jakllsch 20191219]
	net80211: Require SSID configuration to associate with an open AP.
		[jakllsch 20191219]
	dhcpcd(8): Import dhcpcd-8.1.4 [roy 20191220]
	uvm: Rewrite the page allocator to be more efficient and somewhat CPU
		topology aware.  Add preliminary NUMA support.  [ad 20191227]
	crunchgen(1): Added the ability to pass variables in the submake
		command line, and removed all command like flags that are
		not needed anymore [christos 20191229]
	uvm: Reduce lock contention by making the maintentance of page
		replacement state more concurrent. [ad 20191231]
	rge(4): Imported driver for Realtek 8125/8125B PCIe 2.5Gb
		Ethernet devices from OpenBSD. [sevan 20200111]

lib/libc/gen/popen.c
/*	$NetBSD: popen.c,v 1.38 2022/04/19 20:32:15 rillig Exp $	*/

/*
 * Copyright (c) 1988, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software written by Ken Arnold and
 * published in UNIX Review, Vol. 6, No. 8.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#if defined(LIBC_SCCS) && !defined(lint)
#if 0
static char sccsid[] = "@(#)popen.c	8.3 (Berkeley) 5/3/95";
#else
__RCSID("$NetBSD: popen.c,v 1.38 2022/04/19 20:32:15 rillig Exp $");
#endif
#endif /* LIBC_SCCS and not lint */

#include "namespace.h"
#include <sys/param.h>
#include <sys/wait.h>
#include <sys/socket.h>

#include <assert.h>
#include <errno.h>
#include <paths.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>

#include "env.h"

#ifdef __weak_alias
__weak_alias(popen,_popen)
__weak_alias(pclose,_pclose)
#endif

static struct pid {
	struct pid *next;
	FILE *fp;
#ifdef _REENTRANT
	int fd;
#endif
	pid_t pid;
} *pidlist; 
	
#ifdef _REENTRANT
static  mutex_t pidlist_mutex = MUTEX_INITIALIZER;
# define MUTEX_LOCK() \
    do { \
	    if (__isthreaded) \
		    mutex_lock(&pidlist_mutex); \
    } while (0)
# define MUTEX_UNLOCK() \
    do { \
	    if (__isthreaded) \
		    mutex_unlock(&pidlist_mutex); \
    } while (0)
#else
# define MUTEX_LOCK() __nothing
# define MUTEX_UNLOCK() __nothing
#endif

static struct pid *
pdes_get(int *pdes, const char **type)
{
	struct pid *cur;
	int flags = strchr(*type, 'e') ? O_CLOEXEC : 0;
	int serrno;

	if (strchr(*type, '+')) {
		int stype = flags ? (SOCK_STREAM | SOCK_CLOEXEC) : SOCK_STREAM;
		*type = "r+";
		if (socketpair(AF_LOCAL, stype, 0, pdes) < 0)
			return NULL;
	} else  {
		*type = strrchr(*type, 'r') ? "r" : "w";
		if (pipe2(pdes, flags) == -1)
			return NULL;
	}

	if ((cur = malloc(sizeof(*cur))) != NULL)
		return cur;
	serrno = errno;
	(void)close(pdes[0]);
	(void)close(pdes[1]);
	errno = serrno;
	return NULL;
}

static void
pdes_child(int *pdes, const char *type)
{
	struct pid *old;

	/* POSIX.2 B.3.2.2 "popen() shall ensure that any streams
	   from previous popen() calls that remain open in the 
	   parent process are closed in the new child process. */
	for (old = pidlist; old; old = old->next)
#ifdef _REENTRANT
		(void)close(old->fd); /* don't allow a flush */
#else
		(void)close(fileno(old->fp)); /* don't allow a flush */
#endif

	if (type[0] == 'r') {
		(void)close(pdes[0]);
		if (pdes[1] != STDOUT_FILENO) {
			(void)dup2(pdes[1], STDOUT_FILENO);
			(void)close(pdes[1]);
		}
		if (type[1] == '+')
			(void)dup2(STDOUT_FILENO, STDIN_FILENO);
	} else {
		(void)close(pdes[1]);
		if (pdes[0] != STDIN_FILENO) {
			(void)dup2(pdes[0], STDIN_FILENO);
			(void)close(pdes[0]);
		}
	}
}

static void
pdes_parent(int *pdes, struct pid *cur, pid_t pid, const char *type)
{
	FILE *iop;

	/* Parent; assume fdopen can't fail. */
	if (*type == 'r') {
		iop = fdopen(pdes[0], type);
#ifdef _REENTRANT
		cur->fd = pdes[0];
#endif
		(void)close(pdes[1]);
	} else {
		iop = fdopen(pdes[1], type);
#ifdef _REENTRANT
		cur->fd = pdes[1];
#endif
		(void)close(pdes[0]);
	}

	/* Link into list of file descriptors. */
	cur->fp = iop;
	cur->pid =  pid;
	cur->next = pidlist;
	pidlist = cur;
}

static void
pdes_error(int *pdes, struct pid *cur)
{
	free(cur);
	(void)close(pdes[0]);
	(void)close(pdes[1]);
}

FILE *
popen(const char *cmd, const char *type)
{
	struct pid *cur;
	int pdes[2], serrno;
	pid_t pid;

	_DIAGASSERT(cmd != NULL);
	_DIAGASSERT(type != NULL);

	if ((cur = pdes_get(pdes, &type)) == NULL)
		return NULL;

	MUTEX_LOCK();
	(void)__readlockenv();
	switch (pid = vfork()) {
	case -1:			/* Error. */
		serrno = errno;
		(void)__unlockenv();
		MUTEX_UNLOCK();
		pdes_error(pdes, cur);
		errno = serrno;
		return NULL;
		/* NOTREACHED */
	case 0:				/* Child. */
		pdes_child(pdes, type);
		execl(_PATH_BSHELL, "sh", "-c", "--", cmd, NULL);
		_exit(127);
		/* NOTREACHED */
	}
	(void)__unlockenv();

	pdes_parent(pdes, cur, pid, type);

	MUTEX_UNLOCK();

	return cur->fp;
}

FILE *
popenve(const char *cmd, char *const *argv, char *const *envp, const char *type)
{
	struct pid *cur;
	int pdes[2], serrno;
	pid_t pid;

	_DIAGASSERT(cmd != NULL);
	_DIAGASSERT(type != NULL);

	if ((cur = pdes_get(pdes, &type)) == NULL)
		return NULL;

	MUTEX_LOCK();
	switch (pid = vfork()) {
	case -1:			/* Error. */
		serrno = errno;
		MUTEX_UNLOCK();
		pdes_error(pdes, cur);
		errno = serrno;
		return NULL;
		/* NOTREACHED */
	case 0:				/* Child. */
		pdes_child(pdes, type);
		execve(cmd, argv, envp);
		_exit(127);
		/* NOTREACHED */
	}

	pdes_parent(pdes, cur, pid, type);

	MUTEX_UNLOCK();

	return cur->fp;
}

/*
 * pclose --
 *	Pclose returns -1 if stream is not associated with a `popened' command,
 *	if already `pclosed', or waitpid returns an error.
 */
int
pclose(FILE *iop)
{
	struct pid *cur, *last;
	int pstat;
	pid_t pid;

	_DIAGASSERT(iop != NULL);

	MUTEX_LOCK();

	/* Find the appropriate file pointer. */
	for (last = NULL, cur = pidlist; cur; last = cur, cur = cur->next)
		if (cur->fp == iop)
			break;
	if (cur == NULL) {
		MUTEX_UNLOCK();
		errno = ESRCH;
		return -1;
	}

	(void)fclose(iop);

	/* Remove the entry from the linked list. */
	if (last == NULL)
		pidlist = cur->next;
	else
		last->next = cur->next;

	MUTEX_UNLOCK();

	do {
		pid = waitpid(cur->pid, &pstat, 0);
	} while (pid == -1 && errno == EINTR);

	free(cur);

	return pid == -1 ? -1 : pstat;
}
